File: 프로그래머스/LV2/코딩테스트 연습/연습문제/줄 서는 방법/줄 서는 방법.py
from math import factorial

def solution(n, k):
    answer = []
    numbers = [x for x in range(1, n+1)]
    number_idx = 0
    units = [factorial(x) for x in range(n)]
    unit_idx = n-1
    cnt = 0

    while unit_idx > 0:
        cnt += units[unit_idx]
        if cnt >= k:
            answer.append(numbers[number_idx])
            numbers.pop(number_idx)
            cnt -= units[unit_idx]
            unit_idx -= 1
            number_idx = 0
        else:
            number_idx += 1
    answer.append(numbers[number_idx])
    return answer

File: 프로그래머스/LV2/코딩테스트 연습/연습문제/줄 서는 방법/줄 서는 방법.cpp
#include <string>
#include <vector>
using namespace std;

long long factorial(long long n)
{
    return n <= 1 ? 1 : n * factorial(n - 1);
}

vector<int> solution(int n, long long k)
{
    vector<int> answer, numbers;
    vector<long long> units;
    int number_idx = 0;
    int unit_idx = n - 1;
    long long cnt = 0;
    for (int x = 1; x <= n; x++)numbers.push_back(x);
    for (int x = 0; x < n; x++)units.push_back(factorial(x));

    while (unit_idx > 0)
    {
        cnt += units[unit_idx];
        if (cnt >= k)
        {
            answer.push_back(numbers[number_idx]);
            numbers.erase(numbers.begin()+number_idx);
            cnt -= units[unit_idx];
            unit_idx--;
            number_idx = 0;
        }
        else number_idx++;
    }
    answer.push_back(numbers[number_idx]);
    return answer;
}